source/agora/node/Ledger.d

```diff
@@ -356,6 +361,74 @@ public class Ledger
         return true;
     }
 
+    /***************************************************************************
+
+        Returns the public keys of the validators that signed the block at
+        the specified height, and and also not classified as 'missing' validator
+
+        Params:
+            height = block height
+
+        Returns: the public keys of the validators that signed the block at
+                 the specified height, and and also not
+                 classified as 'missing' validator
+
+    ***************************************************************************/
+
+    private Point[] getPKeysOfSignedNonMissingValidators (Height height) nothrow @safe
+    {
+        Point[] keys;
+        Block block;
+        try
+        {
+            block = this.storage.readBlock(height);
+        }
+        catch (Exception e) {
+            log.error("unable to read block at height: {}", height);
+            return null;
+        }
+
+        immutable validator_cnt = enroll_man.getCountOfValidators(height);
+        foreach (i; 0 .. validator_cnt)
+        {
+            auto validator_pkey = enroll_man.getValidatorAtIndex(height, i);
+            if (block.header.validators[i] && !block.header.missing_validators.canFind(i))
```

Comment:
The second check will never be false, because a validator that signed had to sign using its pre-image, hence there's no way to be missing yet sign. If you are missing, you get slashed.

Reply:
I can change, but we have a slashing policy, so what happens if in the future the slashing policy will not be based on pre-images or the signature will not be based pre-images?